File: Javascript/TargetSum.js
// Leetcode Problem - 494

var findTargetSumWays = function(nums,target){
    const memo = new Map(); // Memoization cache

    function backtracking(i, currSum) {
        if (i === nums.length) {
        if (currSum === target) {
            return 1;
        } else {
            return 0;
        }
        }

        if (memo.has(`${i},${currSum}`)) {
        return memo.get(`${i},${currSum}`);
        }

        const positiveCount = backtracking(i + 1, currSum + nums[i]);
        const negativeCount = backtracking(i + 1, currSum - nums[i]);

        const totalCount = positiveCount + negativeCount;
        memo.set(`${i},${currSum}`, totalCount);

        return totalCount;
    }

    return backtracking(0, 0);
}

File: Python/TargetSum.py
##Leetcode Problem - 494

def findTargetSumWays(nums,target):

    memo = {}

    def backtracking(i, curr_sum):
        if i == len(nums):
            if curr_sum == target:
                return 1
            else:
                return 0

        if (i, curr_sum) in memo:
            return memo[(i, curr_sum)]

        positive_count = backtracking(i + 1, curr_sum + nums[i])
        negative_count = backtracking(i + 1, curr_sum - nums[i])

        memo[(i, curr_sum)] = positive_count + negative_count
        return memo[(i, curr_sum)]

    return backtracking(0, 0)


nums= [1000]
target = -1000

print(findTargetSumWays(nums,target))